File: lec06/04_huffman_code/start/huffmanCode.c
// huffmanCode.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// TODO: 허프만 트리 노드 구조체를 정의하세요
typedef struct Node {
  char character;
  int frequency;
  struct Node* left;
  struct Node* right;
} Node;

// TODO: 우선순위 큐를 위한 최소 힙 구조체를 정의하세요
typedef struct MinHeap {
  int size;
  int capacity;
  Node** array;
} MinHeap;

// TODO: 새 노드 생성 함수를 구현하세요
Node* createNode(char character, int frequency) {
  // 새 노드를 할당하고 초기화
  return NULL;  // TODO: 실제 구현으로 교체
}

// TODO: 최소 힙 생성 함수를 구현하세요
MinHeap* createMinHeap(int capacity) {
  // 최소 힙을 생성하고 초기화
  return NULL;  // TODO: 실제 구현으로 교체
}

// TODO: 두 노드 교환 함수를 구현하세요
void swapNode(Node** a, Node** b) {
  // 두 노드의 포인터를 교환
}

// TODO: 최소 힙화 함수를 구현하세요
void minHeapify(MinHeap* minHeap, int index) {
  // 주어진 인덱스부터 힙 속성을 유지하도록 조정
}

// TODO: 최소값 추출 함수를 구현하세요
Node* extractMin(MinHeap* minHeap) {
  // 힙에서 최소값(루트)을 추출하고 힙 속성 유지
  return NULL;  // TODO: 실제 구현으로 교체
}

// TODO: 힙에 노드 삽입 함수를 구현하세요
void insertMinHeap(MinHeap* minHeap, Node* node) {
  // 새 노드를 힙에 삽입하고 힙 속성 유지
}

// TODO: 허프만 트리 구축 함수를 구현하세요
Node* buildHuffmanTree(char characters[], int frequencies[], int size) {
  // 1. 최소 힙 생성
  // 2. 각 문자에 대해 노드를 생성하여 힙에 삽입
  // 3. 허프만 트리 구성 (두 개씩 합치기)
  // 4. 루트 노드 반환
  return NULL;  // TODO: 실제 구현으로 교체
}

// TODO: 허프만 코드 생성 및 출력 함수를 구현하세요
void generateCodes(Node* root, char* code, int depth) {
  // 재귀적으로 트리를 순회하며 각 문자의 코드 생성
  // 왼쪽: '0', 오른쪽: '1'
  // 리프 노드에서 코드 출력
}

// TODO: 허프만 코딩 메인 함수를 구현하세요
void huffmanCoding(char characters[], int frequencies[], int size) {
  // 1. 허프만 트리 구축
  // 2. 허프만 코드 생성 및 출력
}

// 사용 예시
int main() {
  // 예제 데이터
  char characters[] = {'A', 'B', 'C', 'D'};
  int frequencies[] = {60, 25, 10, 5};
  int size = 4;

  printf("입력 데이터:\n");
  for (int i = 0; i < size; i++) {
    printf("문자: %c, 빈도: %d\n", characters[i], frequencies[i]);
  }
  printf("\n");

  // TODO: 허프만 코딩 수행
  // huffmanCoding(characters, frequencies, size);

  return 0;
}

File: lec06/04_huffman_code/start/huffmanCode.py
# huffmanCode.py
import heapq
from collections import defaultdict

# TODO: 허프만 트리 노드 클래스를 구현하세요


class Node:
  """허프만 트리 노드 클래스"""

  def __init__(self, character=None, frequency=0, left=None, right=None):
    # TODO: 노드 초기화
    pass

  # TODO: 우선순위 큐에서 사용하기 위한 비교 연산자를 구현하세요
  def __lt__(self, other):
    # 빈도를 기준으로 비교
    pass

  def __eq__(self, other):
    # 빈도가 같은지 비교
    pass

# TODO: 허프만 트리를 구축하는 함수를 구현하세요


def build_huffman_tree(characters, frequencies):
  """허프만 트리를 구축하는 함수"""
  # 1. 우선순위 큐(최소 힙) 생성
  # 2. 각 문자에 대해 노드 생성하여 힙에 삽입
  # 3. 허프만 트리 구성 (두 개씩 합치기)
  # 4. 루트 노드 반환
  return None

# TODO: 허프만 코드를 생성하는 함수를 구현하세요


def generate_codes(root, code="", codes=None):
  """허프만 코드를 생성하는 함수"""
  # 재귀적으로 트리를 순회하며 각 문자의 코드 생성
  # 왼쪽: '0', 오른쪽: '1'
  # 리프 노드에서 코드 저장
  if codes is None:
    codes = {}

  # TODO: 구현

  return codes

# TODO: 허프만 코딩 메인 함수를 구현하세요


def huffman_coding(characters, frequencies):
  """허프만 코딩 메인 함수"""
  # 1. 허프만 트리 구축
  # 2. 허프만 코드 생성
  # 3. 결과 출력

  print("=== 허프만 코딩 결과 ===")
  # TODO: 구현

  return {}, None  # TODO: 실제 codes와 root 반환

# 트리 구조 출력 함수 (참고용)


def print_tree(root, depth=0, prefix="루트: "):
  """트리 구조를 시각적으로 출력하는 함수"""
  if root is not None:
    print("  " * depth + prefix, end="")
    if root.character is None:
      print(f"내부노드 (빈도: {root.frequency})")
    else:
      print(f"리프노드: {root.character} (빈도: {root.frequency})")

    if root.left or root.right:
      if root.left:
        print_tree(root.left, depth + 1, "L--- ")
      if root.right:
        print_tree(root.right, depth + 1, "R--- ")

# TODO: 압축률을 계산하는 함수를 구현하세요


def calculate_compression_ratio(characters, frequencies, codes):
  """압축률을 계산하는 함수"""
  # 원본 크기와 허프만 코드 크기를 계산하여 압축률 산출
  pass


# 사용 예시
if __name__ == "__main__":
  # 예제 데이터
  characters = ['A', 'B', 'C', 'D']
  frequencies = [60, 25, 10, 5]

  print("입력 데이터:")
  for i, char in enumerate(characters):
    print(f"문자: {char}, 빈도: {frequencies[i]}")
  print()

  # TODO: 허프만 코딩 수행
  # codes, root = huffman_coding(characters, frequencies)

  print("TODO: 허프만 코딩을 구현하세요!")
